Expose HTMLInputElement methods

Reijix opened the pull request:
Hiya!
For my personal project I need some methods of HTMLInputElement, namely
- https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/select
- https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setSelectionRange

I started implementing `select` by copying the existing implementation of `focus`. The only difference is that I used `document.querySelector()` instead of `document.getElementById()`, since the result of the former can be casted to `HTMLInputElement`, while the latter always returns `HTMLElement`.

Next, I'll do the same for `setSelectionRange`, but first is there any feedback? Are these methods even welcome in miso?

dmjio commented on ts/miso/util.ts:
```diff
@@ -12,6 +12,14 @@ export function callFocus(id: string, delay: number): void {
+export function callSelect(id: string, delay: number): void {
+  var setSelect = function () {
+    var e : HTMLInputElement = document.querySelector('#' + id);
+    if (e && e.select) e.select();
```
As @jhrcek mentions, `getElementById` will be a faster lookup. The TypeScript type hierarchy is very rigid for DOM elements. We should be able to `as`-cast into `HTMLInputElement` and use a dynamic property check to avoid issues. Below should act as a safe runtime check, but also satisfy the typechecker.

```typescript
if (e && typeof e['select'] === 'function') (e as HTMLInputElement).select();
```

Also note the `#` prefix will need to be dropped above when using `getElementById`.

dmjio commented on src/Miso/FFI/Internal.hs:
```diff
@@ -508,6 +508,16 @@ focus x = do
 --
+-- Analogous to @document.querySelector('#' + id).select()@.
+select :: MisoString -> JSM ()
+select x = do
```
This will need to be exported.

dmjio commented on src/Miso/FFI/Internal.hs:
```diff
@@ -508,6 +508,16 @@ focus x = do
+-----------------------------------------------------------------------------
+-- | Fails silently if the element is not found.
+--
 -- Analogous to @document.getElementById(id).blur()@
```
Can we flip these two comments, so the one you added is below `Analogous to @document.getElementById(id).blur()@`.

dmjio reviewed (changes requested):
This is a welcome addition, functions that live on DOM elements are awkward to deal with, like `getBoundingClientRect()`, `focus()`, `blur()` etc. 

Also, note that you can invoke `select()` on a `"change"` event for an `<input>` if you use with `with*` variants of event handlers.

https://haddocks.haskell-miso.org/miso/Miso-Html-Event.html#v:onChangeWith

```haskell
-- contrived example

data Action = Changed MisoString DOMRef | Process MisoString

update = \case
  Process selected -> io_ $ do
    consoleLog selected
  Changed newText domRef -> do
    io $ do 
       txt :: MisoString <- fromJSValUnchecked =<< domRef # "select" $ ()
       pure (Process txt)
       
myView :: View m Action
myView = input_ [ onChangeWith Changed ]       
```

This will avoid the `getElementById` lookup, but only works if you're interested in calling `select()` in response to `<input>` value change events.

Reijix commented:
Thanks for the feedback!
I've implemented all of it and have now added `setSelectionRange`, I decided to always set the optional parameter `selectionDirection` to the default `none`. This could of course also be exposed, however you prefer it.

dmjio commented:
The defaults are fine imo, we can generalize on an as-needed basis if need be. 

Thank you for your contribution 🍜 @Reijix
